### PR TITLE
[pnnl/SHAD#93] buffered_insert_iterator

### PR DESCRIPTION
**Pull request by droccom**

### CODE REVIEW

File: include/shad/core/unordered_map.h [hunk 1]
```diff
@@ -50,6 +50,9 @@ template <class Key, class T, class Hash = shad::hash<Key>>
 class unordered_map {
   using hashmap_t = Hashmap<Key, T>;
 
+  template <typename T_>
+  friend class buffered_insert_iterator;
```
Comment (mminutoli) [inline include/shad/core/unordered_map.h]: This says that `buffered_insert_iterator<shad::array>` is a friend of `shad::unordered_map`.

This line should read as:
```C++
friend class buffered_insert_iterator<hashmap_t>;
```

File: include/shad/core/unordered_set.h [hunk 2]
```diff
@@ -49,6 +49,9 @@ template <class Key, class Hash = shad::hash<Key>>
 class unordered_set {
   using set_t = Set<Key>;
 
+  template <typename T>
+  friend class buffered_insert_iterator;
```
Comment (mminutoli) [inline include/shad/core/unordered_set.h]: Same comment applies here.  You want:
```C++
friend class buffered_insert_iterator<set_t>;
```

File: test/unit_tests/stl/common.h [hunk 3]
```diff
@@ -99,8 +100,10 @@ struct create_set_<shad::unordered_set<U>, even> {
   using T = shad::unordered_set<U>;
   std::shared_ptr<T> operator()(size_t size) {
     auto res = std::make_shared<T>(size);
-    auto res_ = res->get();
-    for (size_t i = 0; i < size; ++i) res_->Insert(2 * i + !even);
+    {
+      shad::buffered_insert_iterator<T> ins(*res, res->end());
+      for (size_t i = 0; i < size; ++i) ins = (2 * i + !even);
```
Comment (mminutoli) [inline test/unit_tests/stl/common.h]: This is neat!  I love it.

File: include/shad/core/iterator.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+  /// @brief Flushes pending insertions to the container.
+  void flush() { container_->buffered_flush(); }
+
+  /// @brief Destructor.
+  ///
+  /// The destruction flushes all pending insertions into the container.
+  ~buffered_insert_iterator() { container_->buffered_flush(); }
+
+  inline buffered_insert_iterator& operator*() { return *this; }
```
[the remark this answers is not in the document]
Comment (droccom) [reply include/shad/core/iterator.h]: Thanks!

**Pull request merged**